### PR TITLE
add fcenet

### PR DESCRIPTION
**Pull request by colawyee**
Thank you for your contribution to the MindOCR repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindcv/blob/main/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation

(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs

(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: configs/det/fcenet/README.md
Comment (SamitHuang) [inline configs/det/fcenet/README.md:1]: 图片显示异常。请用拖拽的方式直接将图片上传至git content server

Comment (SamitHuang) [reply configs/det/fcenet/README.md:1]: 下同

Comment (colawyee) [reply configs/det/fcenet/README.md:1]: 已完成

File: requirements.txt [hunk 1]
```diff
@@ -14,3 +14,5 @@ xml-python>=0.4.3
 packaging
 Cython
 lanms>=1.0.2
+imgaug
+Polygon3==3.0.9.1
```
Comment (zhtmike) [inline requirements.txt:18]: what is the difference of Polygon3 and `shapely.geometry.Polygon`，is is possible to use the later one without add new dependency?

File: configs/det/fcenet/README.md [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+|---------------------|----------------|---------------|------------|------------|---------------|-------------|--------------|-----------|-------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| FCENet               | D910x4-MS2.0-F | ResNet50   | ImageNet       | 81.5%     | 86.9%        | 84.1%      | 33 s/epoch   | 7 img/s      | [yaml](fce_icdar15.yaml) | [ckpt](https://download.mindspore.cn/toolkits/mindocr/fcenet/fcenet_resnet50-43857f7f.ckpt) \| [mindir](https://download.mindspore.cn/toolkits/mindocr/fcenet/fcenet_resnet50-43857f7f-5e765611.mindir) |
+
+</div>
+
+#### Notes
+- Context: Training context denoted as {device}x{pieces}-{MS version}{MS mode}, where mindspore mode can be G - graph mode or F - pynative mode with ms function. For example, D910x8-G is for training on 8 pieces of Ascend 910 NPU using graph mode.
+- Note that the training time of FCENet is highly affected by data processing and varies on different machines.
+
```
Comment (Songyuanwei) [inline configs/det/fcenet/README.md:61]: 提供了mindir连接的话，需要在notes里写明mindir的input_shape,参考其他网络

File: configs/det/fcenet/README.md [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+<div align="center">
+
+| **Model**              | **Context**       | **Backbone**      | **Pretrained** | **Recall** | **Precision** | **F-score** | **Train T.**     | **Throughput**   | **Recipe**                            | **Download**                                                                                                                                                                                                |
+|---------------------|----------------|---------------|------------|------------|---------------|-------------|--------------|-----------|-------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| FCENet               | D910x4-MS2.0-F | ResNet50   | ImageNet       | 81.5%     | 86.9%        | 84.1%      | 33 s/epoch   | 7 img/s      | [yaml](fce_icdar15.yaml) | [ckpt](https://download.mindspore.cn/toolkits/mindocr/fcenet/fcenet_resnet50-43857f7f.ckpt) \| [mindir](https://download.mindspore.cn/toolkits/mindocr/fcenet/fcenet_resnet50-43857f7f-5e765611.mindir) |
```
Comment (Songyuanwei) [inline configs/det/fcenet/README.md:54]: 保留小数点后两位

File: mindocr/postprocess/det_fce_postprocess.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+            poly = np.array(boundary[:-1]).reshape(-1, 2).astype(np.float32)
+            score = boundary[-1]
+            points = cv2.boxPoints(cv2.minAreaRect(poly))
+            points = np.int0(points)
+            new_boundaries.append(points.reshape(-1).tolist() + [score])
+
+    return boundaries
+
+
+class FCEPostProcess(object):
```
Comment (liangxhao) [inline mindocr/postprocess/det_fce_postprocess.py:271]: 继承基类 DetBasePostprocess
命令风格和其它类保持一致：FCEPostProcess -> FCEPostprocess

**Comment by zhtmike** [comment]
请在`mindocr/models`添加对应的文件，参考`mindocr/models/det_dbnet.py`，添加url 并在 `mindocr/models/__init__.py` import，以确保CI 会跑相对应的网络构建

File: configs/det/fcenet/README.md [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+
+<!--- Guideline: use url linked to abstract in ArXiv instead of PDF for fast loading.  -->
+
+> FCENet: [Fourier Contour Embedding for Arbitrary-Shaped Text Detection](https://arxiv.org/pdf/2104.10442.pdf)
+
+## 1. Introduction
+
+### FCENet
+
+FCENet is a segmentation-based text detection algorithm. In the text detection scene, algorithms based on segmentation are becoming increasingly popular as they can accurately describe arbitrary shapes of text, including highly-curved text. One of the key highlights of FCENet is its excellent performance on arbitrary-shaped text scenes, which is achieved through deformable convolutions [1] and Fourier transform techniques. Additionally, FCENet possesses advantages of simple post-processing and high generalization, allowing it to achieve good results even with limited training data.
```
[the remark this answers is not in the document]
Comment (colawyee) [reply configs/det/fcenet/README.md]: 已完成

File: mindocr/postprocess/det_fce_postprocess.py [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        fourier_degree=5,
+        num_reconstr_points=50,
+        decoding_type="fcenet",
+        score_thr=0.3,
+        nms_thr=0.1,
+        alpha=1.0,
+        beta=1.0,
+        text_repr_type="poly",
+    ):
+        self.scales = scales
```
[the remark this answers is not in the document]
Comment (colawyee) [reply mindocr/postprocess/det_fce_postprocess.py:288]: 已完成，谢谢

File: mindocr/data/transforms/det_fce_transforms.py [hunk 7]
```diff
@@ -0,0 +1,1461 @@
...
+                results[key] = np.array(valid_masks_list)
+
+        return results
+
+    def __repr__(self):
+        repr_str = self.__class__.__name__
+        return repr_str
+
+
+class RandomRotatePolyInstances:
```
Comment (hadipash) [inline mindocr/data/transforms/det_fce_transforms.py:623]: Replace it with `RandomRotate`.

File: mindocr/models/heads/det_fce_head.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        self.nms_thr = nms_thr
+        self.alpha = alpha
+        self.beta = beta
+        self.text_repr_type = text_repr_type
+        self.train_cfg = train_cfg
+        self.test_cfg = test_cfg
+        self.out_channels_cls = 4
+        self.out_channels_reg = (2 * self.fourier_degree + 1) * 2
+        self.softmax = nn.Softmax(axis=1)
+        self.training = mode
```
Comment (hadipash) [inline mindocr/models/heads/det_fce_head.py:72]: A lot of class members are initialized here but seems not used anywhere.

File: mindocr/models/necks/fpn.py [hunk 9]
```diff
@@ -76,6 +76,56 @@ def _bn(channels, momentum=0.1):
     return nn.BatchNorm2d(channels, momentum=momentum)
 
 
+def Xavier_conv(in_channels, out_channels, kernel_size=3, stride=1, padding=0, pad_mode='same', has_bias=False):
+    init_value = XavierUniform()
+    return nn.Conv2d(in_channels, out_channels,
+                     kernel_size=kernel_size, stride=stride, padding=padding,
+                     pad_mode=pad_mode, weight_init=init_value, has_bias=has_bias)
```
Comment (hadipash) [inline mindocr/models/necks/fpn.py:83]: Make it a member of FCEFPN because it's FCENet specific layer.

File: mindocr/models/necks/fpn.py [hunk 10]
```diff
@@ -76,6 +76,56 @@ def _bn(channels, momentum=0.1):
...
+
+        self.smooth_conv_p3 = Xavier_conv(out_channel, out_channel, kernel_size=3, padding=1, pad_mode='pad',
+                                          has_bias=True)
+
+        self.out_channels = out_channel
+
+    def construct(self, features):
+        c3 = features[1]
+        c4 = features[2]
+        c5 = features[3]
```
Comment (hadipash) [inline mindocr/models/necks/fpn.py:111]: _, c3, c4, c5 = features

File: mindocr/scheduler/dynamic_lr.py [hunk 11]
```diff
@@ -22,6 +22,16 @@
 import math
 from bisect import bisect_right
 
+from mindspore import nn
+
+
+def dynamic_lr(lr, end_lr, epochs, step_per_epoch, by_epoch=True):
```
Comment (hadipash) [inline mindocr/scheduler/dynamic_lr.py:28]: we already have `polynomial_decay` in `mindocr/scheduler/scheduler_factory.py`. Isn't it the same?

**Review by hadipash — CHANGES_REQUESTED** [review]

File: mindocr/data/transforms/det_fce_transforms.py [hunk 12]
```diff
@@ -0,0 +1,1461 @@
...
+        if len(boxes) == 0:
+            return None
+        boxes = self.expand_points_num(boxes)
+        # boxes = np.array(boxes, dtype=np.float32)
+        # txt_tags = np.array(txt_tags, dtype=np.bool)
+        data["polys"] = np.array(boxes, dtype=np.float32)
+        data["texts"] = txts
+        data["ignore_tags"] = txt_tags
+        data["polygons"] = np.array([boxes[i] for i in range(len(txt_tags)) if not txt_tags[i]], dtype=np.float32)
+        data["ignore_polygons"] = np.array([boxes[i] for i in range(len(txt_tags)) if txt_tags[i]], dtype=np.float32)
```
Comment (hadipash) [inline mindocr/data/transforms/det_fce_transforms.py:171]: Why do you split polygons to `polygons` and `ignore_polygons` if you have `ignore_tags`? Anyway, during the augmentation stage need to apply transformations to all polygons, isn't it?

File: configs/det/fcenet/fce_icdar15.yaml [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+    sample_ratio: 1.0
+    transform_pipeline:
+      - DecodeImage:
+          img_mode: RGB
+          to_float32: False
+      - DetLabelEncode:
+      - RandomScale:
+          scale_range: [ 0.75,2.5 ]
+          p: 1.0
+      - DetResize:
```
Comment (hadipash) [inline configs/det/fcenet/fce_icdar15.yaml:72]: `DetResize` and `RandomScale` together will always scale images to 800x800px, which doesn't add variability to training. Previously used `RandomScaling` was doing scaling based on the predefined size of 800px, i.e. an image will be scaled to the following size ranges: [0.75 * 800, 2.5 * 800]. We can simply adjust `scale_range` based on a dataset or add the size limits to `RandomScale` in addition to `scale_range` to avoid an image getting too small or too big.

File: configs/det/fcenet/fce_icdar15.yaml [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+          keep_ratio: True
+          padding: True
+      - RandomCropWithBBox:
+          max_tries: 10
+          min_crop_ratio: 0.1
+          crop_size: [ 640, 640 ]
+          p: 1.0
+      - RandomHorizontalFlip:
+          p: 0.5
+      - RandomRotate:
```
Comment (hadipash) [inline configs/det/fcenet/fce_icdar15.yaml:83]: It is better for `RandomHorizontalFlip` and `RandomRotate` to be applied before `RandomCropWithBBox`. There's a problem with the augmentations order that I am currently trying to fix.

File: mindocr/losses/det_loss.py [hunk 15]
```diff
@@ -157,6 +159,215 @@ def construct(self, pred, gt, mask):
...
+            pass
+            # print(path)
+        # results = dict(
+        #     all_loss = all_loss,
+        #     loss_text=loss_tr,
+        #     loss_center=loss_tcl,
+        #     loss_reg_x=loss_reg_x,
+        #     loss_reg_y=loss_reg_y,
+        # )
+        print(
```
Comment (hadipash) [inline mindocr/losses/det_loss.py:231]: Please update all you print statements to logger as in #429.

File: mindocr/models/necks/fpn.py [hunk 16]
```diff
@@ -76,6 +76,54 @@ def _bn(channels, momentum=0.1):
...
 
+class FCEFPN(nn.Cell):
+    def Xavier_conv(self, in_channels, out_channels, kernel_size=3, stride=1, padding=0, pad_mode='same',
+                    has_bias=False):
+        init_value = XavierUniform()
+        return nn.Conv2d(in_channels, out_channels,
+                         kernel_size=kernel_size, stride=stride, padding=padding,
+                         pad_mode=pad_mode, weight_init=init_value, has_bias=has_bias)
+
+    def __init__(self, in_channels, out_channel):
```
Comment (hadipash) [inline mindocr/models/necks/fpn.py:87]: usually `__init__` defined first.

**Review by hadipash — APPROVED** [review]

**Review by Songyuanwei — APPROVED** [review]
